### PR TITLE
heap sort article added

### PR DESCRIPTION
**Pull request by tk315**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

yes

### Description

Heapsort article added 

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

#48

### CODE REVIEW

**Comment by tk315** [comment]
Please review @ankitaggarwal23 @HarshCasper @iamrajiv

File: docs/sorting-algorithms/heap-sort.mdx [hunk 1]
```diff
@@ -0,0 +1,326 @@
+---
+title: Heap Sort
+slug: /sorting-algorithms/heap-sort
+---
+
+# Introduction
+
+Heap Sort is a sorting algorithm which extracts the maximum or minimum element from the array and swap it with the last element of the array for sorting in ascending or descending order N times.
+It is similar to Selection Sort, the only difference is that it uses Heap data structure for finding minimum or maximum element.
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:9]: Can you give a more detailed introduction here?

File: docs/sorting-algorithms/heap-sort.mdx [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:10]: Can you add an explanation for the algorithm?

File: docs/sorting-algorithms/heap-sort.mdx [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+In Heap Sort, we follow the following algorithm:
+
+* First we will convert array into MaxHeap
+
+* Swap the root of the MaxHeap with the last element of the MaxHeap 
+
+* Reduce the size of MaxHeap by 1 from the last and call Heapify on the root.
+
+* Perform step 2 and step 3 N-1 times
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:21]: Can you explain these steps better. They don't look quite intuitive to me

File: docs/sorting-algorithms/heap-sort.mdx [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+ Swaping first and last element of Heap and reducing the size by 1
+ | 2 | 1 || 4 | 7 |
+
+ Calling Heapify on root
+ | 2 | 1 || 4 | 7 |
+
+ Swaping first and last element of Heap and reducing the size by 1
+ | 1 || 2 | 4 | 7 |
+```
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:47]: Can you use a more intuitive example, please?

File: docs/sorting-algorithms/heap-sort.mdx [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+        {
+            break;
+        }
+        
+    }
+}
+
+int main()
+{
+    int A[8]={0,30,25,40,10,50,20,5};
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:122]: Can you have dynamic input here? I understand that the code is from NeoAlgo but we want to be consistent all across.

File: docs/sorting-algorithms/heap-sort.mdx [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+30
+50
+10
+20
+40
+```
+### Output :
+```
+[10, 20, 30, 40, 50]
+```
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:307]: Use consistent console statements across all the languages

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: docs/sorting-algorithms/heap-sort.mdx (hunk 1 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 2 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx:24]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 3 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 4 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 5 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: I have changed the entire code of heap sort in cpp

File: docs/sorting-algorithms/heap-sort.mdx (hunk 6 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

**Comment by tk315** [comment]
Please review @ankitaggarwal23 @HarshCasper @iamrajiv

File: docs/sorting-algorithms/heap-sort.mdx [hunk 7]
```diff
@@ -0,0 +1,234 @@
+---
+title: Heap Sort
+slug: /sorting-algorithms/heap-sort
+---
+
+# Introduction
+
+Heap Sort is a sorting algorithm which extracts the maximum or minimum element from the array and swap it with the last element of the array 
+for sorting in ascending or descending order N times.It is similar to Selection Sort, the only difference is that it uses Heap data structure 
+for finding minimum or maximum element. It is a In-Place and Non-Stable Sorting Algorithm.
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:10]: Can you wrap it around a single line? Also use `N` in backticks

File: docs/sorting-algorithms/heap-sort.mdx [hunk 8]
```diff
@@ -0,0 +1,234 @@
+---
+title: Heap Sort
+slug: /sorting-algorithms/heap-sort
+---
+
+# Introduction
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:6]: Use Heading-2

File: docs/sorting-algorithms/heap-sort.mdx [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+Heap Sort is a sorting algorithm which extracts the maximum or minimum element from the array and swap it with the last element of the array 
+for sorting in ascending or descending order N times.It is similar to Selection Sort, the only difference is that it uses Heap data structure
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:9]: Keep a space after `.`

Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:9]: `It is similar to the Selection Sort, with the only difference...`

File: docs/sorting-algorithms/heap-sort.mdx [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+slug: /sorting-algorithms/heap-sort
+---
+
+# Introduction
+
+Heap Sort is a sorting algorithm which extracts the maximum or minimum element from the array and swap it with the last element of the array 
+for sorting in ascending or descending order N times.It is similar to Selection Sort, the only difference is that it uses Heap data structure 
+for finding minimum or maximum element. It is a In-Place and Non-Stable Sorting Algorithm.
+
+- Binary Heap: It is a complete binary tree represented using array.There are two types of Binary Heap:
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:12]: Put **Binary Heap** in capital. Keep a space after '.'

File: docs/sorting-algorithms/heap-sort.mdx [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+
+    - **MaxHeap**: Both left and right subtree of every node contains smaller elements
+    - **MinHeap**: Both left and right subtree of every node contains greater elements
+
+- Formulas:
+
+  - **Left Child**:     2*i+1
+  - **Right Child**:    2*i+2
+  - **Parent**:         floor((i-1)/2)
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:21]: Keep the values in backticks

File: docs/sorting-algorithms/heap-sort.mdx [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+
+`i` is the number of elements in the array.
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:23]: `Here i is the number....`


Keep `i` in backticks

File: docs/sorting-algorithms/heap-sort.mdx [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+
+# Algorithm
+
+Heap Sort for Sorting in Increasing Order:
+
+* First we will convert array into MaxHeap by calling Heapify function on the range `[(n-2)/2, 0]`, `(n-2)/2` is the last internal node
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:29]: . Put a full stop after `[(n-2)/2, 0]`

File: docs/sorting-algorithms/heap-sort.mdx [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+
+* Swap the root of the MaxHeap with the last element of the MaxHeap 
+
+* Reduce the size of MaxHeap by 1 from the last and call Heapify on the root.
+
+* Perform step 2 and step 3 `N-1` times
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:35]: `Perform the Step 2 and Step 3`

File: docs/sorting-algorithms/heap-sort.mdx [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+
+Heap Sort for Sorting in Decreasing Order:
+
+* First we will convert array into MinHeap by calling Heapify function on the range `[(n-2)/2, 0]`, `(n-2)/2` is the last internal node
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:39]: Put a full stop after `[(n-2)/2, 0]`

File: docs/sorting-algorithms/heap-sort.mdx [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+
+* Swap the root of the MinHeap with the last element of the MinHeap 
+
+* Reduce the size of MinHeap by 1 from the last and call Heapify on the root.
+
+* Perform step 2 and step 3 `N-1` times
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:45]: `Perform the Step 2 and Step 3`

File: docs/sorting-algorithms/heap-sort.mdx [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+10
+20
+40
+```
+### Output :
+```
+10 20 30 40 50
+```
+
+# Complexity Analysis
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:221]: Keep Heading-2

File: docs/sorting-algorithms/heap-sort.mdx [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+    7(0)                       1(0)                       4(0)                      2(0)                      2(0)                      1(0)                     1(0)
+   /    \                     /    \                     /    \                    /    \                    /    \                    /    \                   /    \
+ 4(1)   2(2)      ->        4(1)   2(2)      ->       1(1)   2(2)      ->       1(1)   4(2)      ->       1(1)   4(2)      ->       2(1)   4(2)      ->      2(1)    4(2)
+ /                          /                          /                         /                         /                         /                        /
+1(3)                      7(3)                       7(3)                      1(3)                      7(3)                      7(3)                     7(3)
+                          
+           Swapping index             Applying Heapify          Swapping index           Applying Heapify           Swapping index            Applying Heapify
+              0 and 3                    on index 0                0 and 2                  on index 0                 0 and 1                   on index 0
+```
+# Code
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:84]: Keep Heading-2

File: docs/sorting-algorithms/heap-sort.mdx [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+
+* First we will convert array into MinHeap by calling Heapify function on the range `[(n-2)/2, 0]`, `(n-2)/2` is the last internal node
+
+* Swap the root of the MinHeap with the last element of the MinHeap 
+
+* Reduce the size of MinHeap by 1 from the last and call Heapify on the root.
+
+* Perform step 2 and step 3 `N-1` times
+
+# Explanation for Heap Sort
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:47]: Keep Heading-2

File: docs/sorting-algorithms/heap-sort.mdx [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:48]: Put some text here if possible

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: docs/sorting-algorithms/heap-sort.mdx (hunk 7 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 8 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 9 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 10 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 11 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 12 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 13 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 14 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 15 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: don

File: docs/sorting-algorithms/heap-sort.mdx (hunk 16 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 17 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 18 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 19 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 20 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx:46]: I have added everything above

**Comment by tk315** [comment]
Please review @HarshCasper

File: docs/sorting-algorithms/heap-sort.mdx [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+    for index in range(length - 1, 0, -1):
+        arr[index], arr[0] = arr[0], arr[index]
+        # then call heapify again with the new array
+        heapify(arr, index, 0)
+ 
+ 
+# Taking Elements to be Sorted
+
+data = []
+n = int(input("Enter the Size of Array : "))
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:187]: The input statements of Python and C++ don't match. Kindly fix them. 

We want the I/O statement to be same for both the languages.

File: docs/sorting-algorithms/heap-sort.mdx [hunk 22]
```diff
@@ -0,0 +1,233 @@
+---
+title: Heap Sort
+slug: /sorting-algorithms/heap-sort
+---
+
+## Introduction
+
+Heap Sort is a sorting algorithm which extracts the maximum or minimum element from the array and swap it with the last element of the array for sorting in ascending or descending order `N` times. It is similar to Selection Sort, with the only difference is that it uses Heap data structure for finding minimum or maximum element. It is a In-Place and Non-Stable Sorting Algorithm.
+
+- BINARY HEAP: It is a complete binary tree represented using array. There are two types of Binary Heap:
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:10]: Keep it `Binary Heap` and bold it

File: docs/sorting-algorithms/heap-sort.mdx [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+
+    - **MaxHeap**: Both left and right subtree of every node contains smaller elements
+    - **MinHeap**: Both left and right subtree of every node contains greater elements
+
+- Formulas:
```
Comment (HarshCasper) [inline docs/sorting-algorithms/heap-sort.mdx:15]: Bold it

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Last set of review done. Sending this to Peer Review now :rocket:

File: docs/sorting-algorithms/heap-sort.mdx (hunk 21 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 22 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 23 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 21 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 22 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

File: docs/sorting-algorithms/heap-sort.mdx (hunk 23 above)
Comment (tk315) [reply docs/sorting-algorithms/heap-sort.mdx]: done

**Comment by tk315** [comment]
I have completely changed the code for heap sort in cpp
It is very much simpler and easier to understand 
Can i make a pr on neoalgo for that?
@HarshCasper

**Comment by HarshCasper** [comment]
> I have completely changed the code for heap sort in cpp
> It is very much simpler and easier to understand
> Can i make a pr on neoalgo for that?
> @HarshCasper

Of Course! Also please fix the merge conflict so that I can merge it.

**Comment by tk315** [comment]
@Harikrishnan6336 Merge Conflict removed

**Comment by tk315** [comment]
@HarshCasper Merge Conflict Removed

**Review by HarshCasper — APPROVED** [review]

**Review by iamrajiv — APPROVED** [review]

**Pull request merged**